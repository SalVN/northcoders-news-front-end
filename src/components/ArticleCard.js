import React from 'react';
import PropTypes from 'prop-types';
import { BrowserRouter as Router, Link } from 'react-router-dom';

const ArticleCard = function (props) {
  return (
    <div className='box'>
      <article className='media'>
        <div className='media-left'>
          {props.number + 1}
        </div>
        <div className='media-left'>
          <button className='button' onClick={props.voteHandler.bind(this, 'up', props.id)}><i className="fa fa-arrow-up" aria-hidden="true"></i></button>
          <div>{props.votes}</div>
          <button className='button' onClick={props.voteHandler.bind(this, 'down', props.id)}><i className="fa fa-arrow-down" aria-hidden="true"></i></button>
        </div>
        <div className='media-content'>
          <div className='content'>
            <Link to={`/articles/${props.id}`}>
              <h3 className='title is-3'>{props.title}</h3>
            </Link>
            <p>
              <span>added by: {props.author}</span>
              <span>tags: {props.tags}</span>
              <span>comments: {props.comment_count}</span>
            </p>
          </div>
        </div>
      </article>
    </div>
  );
};

ArticleCard.propTypes = {
  title: PropTypes.string.isRequired,
  votes: PropTypes.number.isRequired,
  number: PropTypes.number.isRequired,
  author: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  tags: PropTypes.string.isRequired,
  comment_count: PropTypes.number.isRequired,
  voteHandler: PropTypes.func.isRequired
};


export default ArticleCard;
