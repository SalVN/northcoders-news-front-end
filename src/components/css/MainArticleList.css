.page-title {
    margin-top: 10px;
    font-weight: bold;
}

.select-menu {
    margin-top: 15px;
    font-size: 14px;
}

.label {
    margin-top: 17px;
    font-size: 14px;
    text-align: right;
}
