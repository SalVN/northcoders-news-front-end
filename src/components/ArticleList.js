import React from 'react';
import ArticleCard from './ArticleCard';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import * as actions from '../actions/actions';

class ArticleList extends React.Component {
  componentDidMount() {
    if (this.props.users && this.props.users.length < 1) {
      this.props.fetchUsers();
    }
  }
  render() {
    const users = this.props.users;
    const voteHandler = this.props.voteArticle;
    return (
      <div id='ArticleList' className='content'>
        {this.props.articles.length > 0 &&
          this.props.articles.map((article, i) => {
            const username = article.created_by;
            let index;
            users &&
              users.forEach((user, i) => {
                if (user.username === username) index = i;
              });
            return (
              <ArticleCard
                key={article._id}
                title={article.title}
                number={i}
                votes={article.votes}
                author={article.created_by}
                id={article._id}
                tags={article.belongs_to}
                comment_count={article.comment_count}
                voteHandler={voteHandler}
                userData={users[index]}
              />
            );
          })
        }
      </div>
    );
  }
}

function mapDispatchToProps(dispatch) {
  return {
    fetchUsers: () => {
      dispatch(actions.fetchUsers());
    }
  };
}

function mapStateToProps(state) {
  return {
    loading: state.users.loading,
    users: state.users.users
  };
}

ArticleList.propTypes = {
  articles: PropTypes.array.isRequired,
  loading: PropTypes.bool.isRequired,
  fetchUsers: PropTypes.func.isRequired,
  voteArticle: PropTypes.func.isRequired,
  users: PropTypes.array.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(ArticleList);