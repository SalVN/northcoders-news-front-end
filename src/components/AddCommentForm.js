import React from 'react';
import PropTypes from 'prop-types';

import './css/AddCommentForm.css';

class AddCommentForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            text: ''
        };
        this.handleChange = this.handleChange.bind(this);
    }
    render() {
        return (
            <div>
                {this.props.showForm
                    ?
                    <div className='box box-style box-style-comment box-style-comment'>
                        <article className='media'>
                            <div className='media-left'>
                                <img className='comment-form-image' src={this.props.user.avatar_url} />
                            </div>
                            <div className='media-content'>
                                <div className='content'>
                                    <form onSubmit={this.props.handleSubmit}>
                                        <label className='comment-form-heading' htmlFor="comment-input">Join the conversation:</label>
                                        <br />
                                        <div className='comment-form' onBlur={this.props.toggleForm.bind(this, this.state.text)}>
                                            <textarea className='add-comment-textarea' onChange={this.handleChange} id="comment-input" defaultValue={this.state.text} placeholder='Join the conversation' />
                                        </div>
                                        <div className='button-div'>
                                            <button className='button add-comment-submit' type='submit'>Submit</button>
                                        </div>
                                    </form>
                                </div>
                            </div>
                        </article>
                    </div>
                    : <div className='box box-style box-style-comment box-style-comment-small'>
                        <article className='media'>
                            <div className='media-left'>
                                <img className='comment-form-image' src={this.props.user.avatar_url} />
                            </div>
                            <div className='media-content'>
                                <div className='content'>
                                    <div onClick={this.props.toggleForm.bind(this, null)} className='comment-form-small'>
                                        <p>Join the conversation</p>
                                    </div>
                                </div>
                            </div>
                        </article>
                    </div>
                }
            </div>
        );
    }
    handleChange(e) {
        this.setState({
            text: e.target.value
        });
    }
}

AddCommentForm.propTypes = {
    id: PropTypes.string.isRequired,
    addComment: PropTypes.func.isRequired,
    toggleForm: PropTypes.func.isRequired,
    user: PropTypes.object.isRequired,
    showForm: PropTypes.bool.isRequired,
    handleSubmit: PropTypes.func.isRequired
};

export default AddCommentForm;