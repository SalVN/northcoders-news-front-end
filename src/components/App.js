import React, {Component} from 'react';
import PropTypes from 'prop-types';

import TopicNavBar from './TopicNavBar';
import MainBanner from './MainBanner';
import UserCard from './UserCard';
import UserListCard from './UserListCard';

import './css/App.css';

export class App extends Component {
  render() {
    let pageStyling = window.innerWidth < 800 ? 'main-page-mobile' : 'main-page';
    return (
      <div className='page'>
        <TopicNavBar />
        <MainBanner />
        <div className={pageStyling}>
          <div className={'columns is-gapless'}>
            <div className='column is-three-quarters'>
              {this.props.children}
            </div>
            <div className='column is-one-quarter'>
              <div className='is-hidden-mobile'>
                <UserCard />
              </div>
              <UserListCard />
            </div>
          </div>
        </div>
      </div>
    );
  }
}

App.propTypes = {
  children: PropTypes.element.isRequired
};

export default App;
