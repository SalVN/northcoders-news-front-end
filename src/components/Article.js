import React from 'react';
import PropTypes from 'prop-types';
import { BrowserRouter as Router, Link } from 'react-router-dom';

import './css/Article.css';

const Article = function (props) {
    return (
        <div className='box'>
            <div className='columns'>
                <div className='column is-1'>
                    <button className='button is-black is-inverted is-small'><i className="fa fa-arrow-up" aria-hidden="true"></i></button>
                    <p className='comment-card-votes'><strong>{props.article.votes}</strong></p>
                    <button className='button is-black is-inverted is-small'><i className="fa fa-arrow-down" aria-hidden="true"></i></button>
                </div>
                <div className='column'>
                    <div className='title is-3 article-title'>{props.article.title}</div>
                    <p>
                        <span>
                            <span className='article-subheading-key'>added by:</span>
                            <Link to='#'>
                                <span>
                                    {props.user &&
                                        <img className='article-card-avatar' src={props.user.avatar_url} alt='user avatar' />
                                    }
                                </span>
                                <span className='article-subheading-value'>
                                    <strong>{props.article.created_by}</strong>
                                </span>
                            </Link>
                        </span>
                        <span>
                            <span className='article-card-subheading-key'>tags:</span>
                            <Link to={`/topics/${props.article.belongs_to}/articles`}>
                                <span className='article-card-subheading-value'>
                                    <strong>{props.article.belongs_to}</strong>
                                </span>
                            </Link>
                        </span>
                        <span>
                            <span className='article-card-subheading-key'>comments:</span>
                            {/*<Link to='#'>*/}
                            <span className='article-card-subheading-value-notlink'>
                                <strong>{props.article.commentCount}</strong>
                            </span>
                            {/*</Link>*/}
                        </span>
                    </p>
                    <div>{props.article.body}</div>
                </div>
            </div>
        </div>
    );
};

Article.propTypes = {
    article: PropTypes.object.isRequired,
    user: PropTypes.object.isRequired
};

export default Article;