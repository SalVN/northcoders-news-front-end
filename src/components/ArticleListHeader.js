import React from 'react';
import PropTypes from 'prop-types';

import './css/MainArticleList.css';

const ArticleListHeader = function (props) {
    return (
        <div className='content'>
            <div className='columns'>
                <div className='column is-3'>
                    <h2 className='page-title'><strong>{props.title}</strong></h2>
                </div>
                <div className='column is-1 is-offset-5 label'>
                    <label htmlFor="Sort by">Sort by:</label>
                </div>
                <div className='column is-3'>
                    <div className='select is-primary select-menu'>
                        <select name="Sort by" onChange={props.handleClickSelect}>
                            <option value="votes">Popularity</option>
                            <option value="comments">Number of Comments</option>
                            <option value="random">Random</option>
                        </select>
                    </div>
                </div>
            </div>
        </div>
    );
};

ArticleListHeader.propTypes = {
    title: PropTypes.string.isRequired,
    handleClickSelect: PropTypes.func.isRequired
};

export default ArticleListHeader;