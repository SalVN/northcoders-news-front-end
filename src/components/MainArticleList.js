import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import * as actions from '../actions/actions';
import ArticleList from './ArticleList';
import ArticleListHeader from './ArticleListHeader';
import './css/MainArticleList.css';

class MainArticleList extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            sortBy: 'votes',
            showDropdown: false,
            maximum: 10
        };
        this.voteHandlerMainArticles = this.voteHandlerMainArticles.bind(this);
        this.handleClickSelect = this.handleClickSelect.bind(this);
        this.toggleDropdown = this.toggleDropdown.bind(this);
        this.viewMoreArticles = this.viewMoreArticles.bind(this);
    }
    componentDidMount() {
        if (this.props.articles && this.props.articles.length < 1) {
            this.props.fetchArticles();
        }
    }
    render() {
        return (
            <div id='ArticleList' className='content'>
                <ArticleListHeader
                    title={'Articles'}
                    handleClickSelect={this.handleClickSelect}
                    toggleDropdown={this.toggleDropdown}
                    showDropdown={this.state.showDropdown}
                />
                <ArticleList
                    articles={this.props.articles}
                    voteArticle={this.voteHandlerMainArticles}
                    sortBy={this.state.sortBy}
                    maximum={this.state.maximum}
                    viewMoreArticles={this.viewMoreArticles}
                />
            </div>
        );
    }
    voteHandlerMainArticles(vote, id) {
        this.props.voteArticle(vote, id);
    }
    handleClickSelect(e) {
        e.preventDefault();
        this.setState({
            sortBy: e.target.attributes[0].value,
            showDropdown: false
        });
    }
    toggleDropdown() {
        this.setState({
            showDropdown: !this.state.showDropdown
        });
    }
    viewMoreArticles () {
        const newMax = this.state.maximum + 10;
        this.setState({
            maximum: newMax
        });
    }
}

function mapDispatchToProps(dispatch) {
    return {
        fetchArticles: () => {
            dispatch(actions.fetchArticles());
        },
        voteArticle: (vote, id) => {
            dispatch(actions.voteArticle(vote, id));
        }
    };
}

function mapStateToProps(state) {
    return {
        articles: state.articles.articles,
        loading: state.articles.loading,
    };
}

MainArticleList.propTypes = {
    articles: PropTypes.array.isRequired,
    loading: PropTypes.bool.isRequired,
    fetchArticles: PropTypes.func.isRequired,
    voteArticle: PropTypes.func.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(MainArticleList);