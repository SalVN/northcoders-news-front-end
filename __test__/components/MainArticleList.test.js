import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { MemoryRouter } from 'react-router-dom';
import ReactShallowRenderer from 'react-test-renderer/shallow';
import sinon from 'sinon';

const mockStore = configureStore();
const initialState = {};

import { MainArticleList } from '../../src/components/MainArticleList';

describe('MainArticleList', () => {
    const articles = [{
        _id: '59b01acf006c8dbca914672f',
        title: 'Football is fun',
        body: 'something',
        belongs_to: 'football',
        __v: 0,
        votes: 3,
        comment_count: 0
    },
    {
        _id: '59b01acf006c8dbca914672e',
        title: 'Cats are great',
        body: 'something',
        belongs_to: 'cats',
        __v: 0,
        votes: 2,
        comment_count: 2
    }];

    it('is a function', () => {
        expect(typeof MainArticleList).toEqual('function');
    });

    it('renders', () => {
        const store = mockStore(initialState);
        const wrapper = shallow(<MainArticleList
            store={store}
            articles={articles}
            articlesLoading={false}
            fetchArticles={x => x}
            fetchUsers={x => x}
            voteArticle={x => x}
        />);
        expect(wrapper.children().length).toEqual(2);
    });

    it('renders correctly', () => {
        const store = mockStore(initialState);
        const renderer = new ReactShallowRenderer();
        const tree = renderer.render(
            <Provider store={store}>
                <MemoryRouter>
                    <MainArticleList
                        store={store}
                        articles={articles}
                        articlesLoading={false}
                        fetchArticles={x => x}
                        fetchUsers={x => x}
                        voteArticle={x => x} />
                </MemoryRouter>
            </Provider>
        );
        expect(tree).toMatchSnapshot();
    });

    it('has an initial state', () => {
        const wrapper = shallow(<MainArticleList
            articles={articles}
            articlesLoading={false}
            fetchArticles={x => x}
            fetchUsers={x => x}
            voteArticle={x => x}
        />);
        expect(wrapper.state('sortBy')).toBe('votes');
        expect(wrapper.state('showDropdown')).toBe(false);
        expect(wrapper.state('maximum')).toBe(10);
        expect(wrapper.state('voted')).toBe(false);
    });

    it('fetches users if the componentWillReceiveProps following a vote', () => {
        const spy = sinon.stub();
        const wrapper = shallow(<MainArticleList
            articles={articles}
            articlesLoading={false}
            fetchArticles={x => x}
            fetchUsers={spy}
            voteArticle={x => x}
        />);
        expect(spy.callCount).toBe(0);
        wrapper.setState({ voted: true });
        wrapper.setProps({ articles: articles });
        expect(spy.callCount).toBe(1);
        expect(wrapper.state('voted')).toBe(false);
        wrapper.setProps({ articles: articles });
        expect(spy.callCount).toBe(1);
    });

    it('displays a loading icon if this.props.articlesLoading', () => {
        const enzymeWrapperA = shallow(<MainArticleList
            articles={articles}
            articlesLoading={false}
            fetchArticles={x => x}
            fetchUsers={x => x}
            voteArticle={x => x}
        />);
        expect(enzymeWrapperA.find('.fa-spin').length).toBe(0);

        const enzymeWrapperB = shallow(<MainArticleList
            articles={articles}
            articlesLoading={true}
            fetchArticles={x => x}
            fetchUsers={x => x}
            voteArticle={x => x}
        />);
        expect(enzymeWrapperB.find('.fa-spin').length).toBe(1);
    });

    it('only displays the ArticleList component if this.props.articlesLoading === false', () => {
        const enzymeWrapperA = shallow(<MainArticleList
            articles={articles}
            articlesLoading={true}
            fetchArticles={x => x}
            fetchUsers={x => x}
            voteArticle={x => x}
        />);

        const enzymeWrapperB = shallow(<MainArticleList
            articles={articles}
            articlesLoading={false}
            fetchArticles={x => x}
            fetchUsers={x => x}
            voteArticle={x => x}
        />);
        expect(enzymeWrapperA.children().nodes[1].type.displayName).toBe(undefined);
        expect(enzymeWrapperB.children().nodes[1].type.displayName).toBe('Connect(ArticleList)');
    });
});