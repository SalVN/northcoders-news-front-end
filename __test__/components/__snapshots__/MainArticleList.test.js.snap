// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainArticleList renders correctly 1`] = `
<MemoryRouter>
  <MainArticleList
    articles={
      Array [
        Object {
          "__v": 0,
          "_id": "59b01acf006c8dbca914672f",
          "belongs_to": "football",
          "body": "something",
          "comment_count": 0,
          "title": "Football is fun",
          "votes": 3,
        },
        Object {
          "__v": 0,
          "_id": "59b01acf006c8dbca914672e",
          "belongs_to": "cats",
          "body": "something",
          "comment_count": 2,
          "title": "Cats are great",
          "votes": 2,
        },
      ]
    }
    articlesLoading={false}
    fetchArticles={[Function]}
    fetchUsers={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    voteArticle={[Function]}
  />
</MemoryRouter>
`;
